pkg/ssh/ssh.go

```diff
@@ -154,52 +154,40 @@ type sshDialer interface {
 	Dial(network, addr string, config *ssh.ClientConfig) (*ssh.Client, error)
 }
 
-// Real implementation of sshDialer
-type realSSHDialer struct{}
-
-var _ sshDialer = &realSSHDialer{}
-
-func (d *realSSHDialer) Dial(network, addr string, config *ssh.ClientConfig) (*ssh.Client, error) {
-	return ssh.Dial(network, addr, config)
-}
-
-// timeoutDialer wraps an sshDialer with a timeout around Dial(). The golang
+// timeoutDialer implements a Dial() method that will timeout. The golang
 // ssh library can hang indefinitely inside the Dial() call (see issue #23835).
-// Wrapping all Dial() calls with a conservative timeout provides safety against
-// getting stuck on that.
 type timeoutDialer struct {
-	dialer  sshDialer
 	timeout time.Duration
 }
 
 // 150 seconds is longer than the underlying default TCP backoff delay (127
 // seconds). This timeout is only intended to catch otherwise uncaught hangs.
 const sshDialTimeout = 150 * time.Second
 
-var realTimeoutDialer sshDialer = &timeoutDialer{&realSSHDialer{}, sshDialTimeout}
+var defaultTimeoutDialer sshDialer = &timeoutDialer{sshDialTimeout}
 
 func (d *timeoutDialer) Dial(network, addr string, config *ssh.ClientConfig) (*ssh.Client, error) {
-	var client *ssh.Client
-	errCh := make(chan error, 1)
-	go func() {
-		defer runtime.HandleCrash()
-		var err error
-		client, err = d.dialer.Dial(network, addr, config)
-		errCh <- err
+	conn, err := net.Dial(network, addr)
+	if err != nil {
+		return nil, err
+	}
+	conn.SetDeadline(time.Now().Add(d.timeout))
```

Comment:
Does conn.SetDeadline break long-lived connections through the proxyDialer? (e.g. exec, attach). To test that, exec into a pod running a shell (e.g. `kubectl exec --namespace=kube-system kube-dns-v9-1234 -c kube2sky`), and leave it alone for > 2.5 minutes. If the shell is still alive, we're okay, but I'm guessing it won't be.